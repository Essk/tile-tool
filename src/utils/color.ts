/* tslint:disable: no-bitwise */

import { FileEntity } from './Storeable';
export class Color extends FileEntity {
    public static asHex(color: IColor ): string {
        return  `#${Color.decToHex(color.red)}${Color.decToHex(color.green)}${Color.decToHex(color.blue)}`;
    }
    public static decToHex(dec: number): string {
        const clampedDec = clampedValue(dec);
        const hex = clampedDec.toString(16).toUpperCase();
        return `${hex}${hex}`.substring(0, 2);
    }
    public static nextColors(): Color[] {
        const color = new Color();
        const colors: Color[] = [];
        const three = color.threeBitScale;
        const two = color.twoBitScale;
        // tslint:disable-next-line:prefer-for-of
        for (let r = 0; r < three.length; r ++) {
            // tslint:disable-next-line:prefer-for-of
            for (let g = 0; g < three.length; g++) {
                // tslint:disable-next-line:prefer-for-of
                for (let b = 0; b < three.length; b++) {
                    colors.push( new Color({ red: three[r], green: three[g], blue: three[b] }));
                }
            }
        }
        return colors;
    }
    public static hydrate(fileColor: ColorFileData) {
        return new Color( fileColor );
    }
    public readonly twoBitScale: number [];
    public readonly threeBitScale: number [];
    private _red: number;
    private _green: number;
    private _blue: number;
    private _hex: string;
    constructor(opts?: IColor | ColorFileData) {
        super();
        this.threeBitScale = [0, 36, 73, 109, 146, 182, 219, 255];
        this.twoBitScale = [0, 85, 170, 255];
        if (typeof opts !== 'undefined' && this.isFromFile(opts)) {
            this._red = opts._red;
            this._green = opts._green;
            this._blue = opts._blue;
        } else if ( typeof opts !== 'undefined' && ! this.isFromFile(opts)) {
            this._red = this.normaliseThree(opts.red);
            this._green = this.normaliseThree(opts.green);
            this._blue = this.normaliseThree(opts.blue);
        } else {
            this._red =  255;
            this._green = 255;
            this._blue =  255;
        }
        this._hex = Color.asHex({red: this._red, green: this._green, blue: this._blue});
    }
    private normaliseThree(value: number): number {
        const clamped = clampedValue(value);
        const eightBitVal = this.roundEightBit(clamped);
        return this.matchNextValue(eightBitVal, this.threeBitScale);
    }
    private normaliseTwo(value: number): number {
        const clamped = clampedValue(value);
        const eightBitVal = this.roundEightBit(clamped);
        return this.matchNextValue(eightBitVal, this.twoBitScale);
    }
    private roundEightBit(num: number) {
        const remainder = num % 32;
        if (remainder !== 0 ) {
            if (remainder >= 16) {
                return num + (32 - remainder) ;
            }
            return num - remainder;
        }
        return num;
    }
    private matchNextValue(num: number, scale: number[]): number {
        let i = 0;
        const values = scale;
        const lastInScale = scale[scale.length - 1];
        if (num >= lastInScale) {
            return lastInScale;
        }
        while ( (num >= values [i] && num >= values [i + 1]) && i < (values.length - 1) ) {
            i++;
        }
        const gapToLower = num - values[i];
        const gapToHigher = values[i + 1] - num;
        if ( gapToLower < gapToHigher) {
            return values[i];
        } else {
            return values[i + 1];
        }
    }
    private isFromFile(opts: IColor | ColorFileData): opts is ColorFileData {
        return (opts as ColorFileData)._red !== undefined;
    }
    get red() {
        return this._red;
    }
    get green() {
        return this._green;
    }
    get blue() {
        return this._blue;
    }
    get hex() {
        return this._hex;
    }

}

export interface IColor {
    red: number;
    green: number;
    blue: number;
}

export type ColorFileData = {
    _red: number;
    _green: number;
    _blue: number;
};

/**
 * makes sure we have an integer 0 - 255
 */
export function clampedValue(value: number): number {
    const int = parseInt( value.toString(10), 10);
    return ( int >= 0 && int <= 255 ) ? int : 255;
}




