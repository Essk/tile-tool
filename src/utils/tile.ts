import uniqid from 'uniqid';

type Opts = {
    name?: string;
    id?: string;
    px?: number[];
    paletteSet?: string;
    currentPalette?: any;
};
type TileFileData = {
    _name: string;
    _id: string;
    _px: number[];
    _paletteSet: string;
    _currentPalette: any;
};
export class Tile {
    private _name: string;
    private _id: string;
    private _px: number[];
    private _len: number;
    private _paletteSet: string;
    private _currentPalette: number;
    constructor(opts: Opts | TileFileData) {
        if (this.isFromFile(opts)) {
            this._name = opts._name;
            this._id = opts._id;
            this._len = 64;
            this._px = opts._px ;
            this._paletteSet = opts._paletteSet;
            this._currentPalette = opts._currentPalette;
        } else {
            this._name = opts.name || 'new tile';
            this._id = opts.id || this.makeId();
            this._len = 64;
            this._px = opts.px ? this.makePx(this._len, opts.px) : this.makePx(this._len) ;
            this._paletteSet = opts.paletteSet || '';
            this._currentPalette =  this.validPalette(opts.currentPalette) ? opts.currentPalette : 0;
        }
    }
    public updatePx(idx: number, value: number) {
        if (this.validIdx(idx) && this.validValue(value)) {
            this.px[idx] = value;
        }
    }
    private makeId(): string {
        return uniqid('tile-');
    }
    private makePx(len: number, px?: number[]): number[] {
        const _px = px || [];
        if (_px.length > 0 && _px.length <= len) {
            for (let i = _px.length; i < len; i++) {
                _px.push(0);
            }
        } else if (_px.length > len) {
            return _px.splice(len);
        }
        return _px;
    }
    get px(): number[] {
        return this._px;
    }
    set px(newPx: number[]) {
        this._px = newPx;
    }
    get name(): string {
        return this._name;
    }
    get id(): string {
        return this._id;
    }
    get paletteSet(): string {
        return this._paletteSet;
    }
    set palette_set(paletteSet: string) {
        this._paletteSet = paletteSet;
    }
    get currentPalette(): number {
        return this._currentPalette;
    }

    set currentPalette(idx: number) {
        this._currentPalette = this.validPalette(idx) ? idx : 0 ;
    }

    private validPalette(idx: number ): boolean {
        if (idx === undefined) {
            return false;
        }
        return Number.isInteger(idx)
        && idx >= 0
        && idx < 16;
    }
    private validIdx(idx: number): boolean {
        return Number.isInteger(idx) && idx >= 0 && idx < this._len;
    }
    private validValue(value: number): boolean {
        return Number.isInteger(value) && value >= 0 && value < 16;
    }
    private isFromFile(opts: Opts | TileFileData): opts is TileFileData {
        return (opts as TileFileData)._id !== undefined;
    }

}
