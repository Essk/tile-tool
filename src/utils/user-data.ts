const electron = require('electron');
const path = require('path');
const fs = require('fs');

type Opts = {
  configName: string;
  defaults: object;
};

type Dict = { [key: string]: any };

export class FileDataStore {
    public data: Dict;
    private path: string;
    constructor(opts: Opts) {
        const userDataPath = (electron.app || electron.remote.app).getPath(
        'userData',
        );
        this.path = path.join(userDataPath, opts.configName + '.json');
        this.data = parseDataFile(this.path, opts.defaults);
        // do a save in case this is first run
        // can probably handle that circumstance
        // more elegantly
        Object.keys(this.data).forEach( (key) => {
          this.set(key, this.data[key]);
        });
    }


  // This will just return the property on the `data` object
  public get(key: string): string {
    return this.data[key];
  }

  // ...and this will set it
  public set(key: string, val: string): void {
    this.data[key] = val;
    // Note that in a real app, we would try/catch this.
    try {
      fs.writeFileSync(this.path, JSON.stringify(this.data));
    } catch (error) {
      console.log('could not save to disk');
      console.error(error);
    }
  }
}

function parseDataFile(filePath: string, defaults: object) {
  // We'll try/catch it in case the file doesn't exist yet, which will be the case on the first application run.
  // `fs.readFileSync` will return a JSON string which we then parse into a Javascript object
  try {
    return JSON.parse(fs.readFileSync(filePath));
  } catch (error) {
    // if there was some kind of error, return the passed in defaults instead.
    return defaults;
  }
}
