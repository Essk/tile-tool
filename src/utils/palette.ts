import { Color } from './color';
import uniqid from 'uniqid';
type Opts = {
  name: string;
};

export class Palette {
  private _name: string;
  private _id: string;
  private _colors: Color[];
  private _len: number;
  constructor(opts: Opts) {
    this._name = opts.name || 'new palette';
    this._id = this.makeId();
    this._len = 16;
    this._colors = this.makeColors(this._len);
  }
  get name(): string {
    return this._name;
  }
  get id(): string {
    return this._id;
  }
  get colors(): Color[] {
    return this._colors;
  }
  set colors(newColors: Color[]) {
    this._colors = newColors;
  }
  public updateColor(idx: number, color: Color) {
    if (this.validIdx(idx)) {
      this.colors.splice(idx, 1, color);
    }
  }
  private makeColors(len: number) {
    const colors = [];
    for (let i = 0; i < len; i++) {
      colors.push(new Color({ red: 255, green: 255, blue: 255 }));
    }
    return colors;
  }
  private makeId(): string {
    return uniqid('palette-');
  }
  private validIdx(idx: number): boolean {
    return Number.isInteger(idx) && idx >= 0 && idx < this._len;
  }
}
