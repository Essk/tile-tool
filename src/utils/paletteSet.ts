import { FileEntity } from './Storeable';
import { Palette, PaletteFileData } from './palette';
import uniqid from 'uniqid';
type Opts = {
  id?: string;
  name?: string;
  palettes?: Palette[];

};
export type PaletteSetFileData = {
  _id: string;
  _name: string;
  _palettes: PaletteFileData[];
};
export class PaletteSet extends FileEntity {
  public static hydrate(filePaletteSet: PaletteSetFileData) {
    return new PaletteSet( filePaletteSet );
  }
  private _name: string;
  private _id: string;
  private _len: number;
  private _palettes: Palette[];
  constructor(opts: Opts | PaletteSetFileData) {
    super();
    this._len = 16;
    if (this.isFromFile(opts)) {
      this._name = opts._name;
      this._id = opts._id;
      this._palettes = opts._palettes.map( (palette) => Palette.hydrate(palette) );
    } else {
      this._name = opts.name || 'new palette';
      this._id = opts.id || this.makeId();
      this._palettes = this.validatePalettes( opts.palettes);
    }
  }
  get name(): string {
    return this._name;
  }
  get id(): string {
    return this._id;
  }
  get palettes() {
    return this._palettes;
  }
  public removePaletteByIndex(idx: any): void {
    if (this.validIdx(idx)) {
      this.palettes.splice(idx, 1);
    }
  }
  public addPalette(palette: any): void {
    if (palette instanceof Palette && this._palettes.length < this._len) {
      this._palettes.push(palette);
    }
  }
  private makeId(): string {
    return uniqid('paletteSet-');
  }
  private validIdx(idx: any): boolean {
    return Number.isInteger(idx) && idx >= 0 && idx < this._len;
  }

  private validatePalettes( palettes: any[] | undefined ): Palette[] {
    let clippedPalettes = palettes;
    if (palettes !== undefined &&  palettes.length > this._len) {
      clippedPalettes = palettes.splice(this._len - 1 );
    }
    if (clippedPalettes !== undefined ) {
      return clippedPalettes.map( (palette) => {
        if ( palette instanceof Palette ) {
          return palette;
        } else {
          return new Palette({});
        }
       } );
    } else  {
      return [];
    }
  }

  private isFromFile(opts: Opts | PaletteSetFileData): opts is PaletteSetFileData {
    return (opts as PaletteSetFileData)._id !== undefined;
  }
}

