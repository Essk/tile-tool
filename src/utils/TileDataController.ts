import { FileDataStore } from './user-data';
import { Tile, TileFileData } from '@/utils/tile';
import { PaletteSet, PaletteSetFileData } from '@/utils/paletteSet';
import { Color } from '@/utils/color';
import { Palette } from './palette';


export class TileDataController {
    private _tiles: Tile[];
    private _paletteSets: PaletteSet[];
    private defaultPaletteSet = new PaletteSet({ name: 'new palette set', palettes: [ new Palette({}) ] });
    private defaultTile = new Tile({ name: 'new tile', paletteSet: this.defaultPaletteSet.id });
      private fileStore = new FileDataStore({
      configName: 'user-data', defaults: {
        tiles: [this.defaultTile],
        paletteSets: [this.defaultPaletteSet],
      },
    });
    constructor() {
        this._tiles = this.fileStore.data.tiles.map( (tile: TileFileData ) =>  Tile.hydrate(tile) );
        this._paletteSets = this.fileStore.data.paletteSets.
        map( (paletteSet: PaletteSetFileData ) =>  PaletteSet.hydrate(paletteSet) );
    }
    get tiles() {
        return this._tiles;
    }
    get paletteSets() {
        return this._paletteSets;
    }
}
