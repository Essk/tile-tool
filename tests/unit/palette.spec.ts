import { Palette } from '@/utils/palette.ts';
import { Color } from '@/utils/color';

describe('Palette', () => {
  it('instantiates', () => {
    const palette = new Palette({
      name: '',
    });
    expect(palette).toBeInstanceOf(Palette);
  });

  it('takes a name from passed options', () => {
    const palette = new Palette({name : 'a unique palette'});
    expect(palette.name).toBe('a unique palette');
  });
  it('optionally accepts an id in opts', () => {
    const palette = new Palette({ id: 'badger_snuff'});
    expect(palette.id).toBe('badger_snuff');
  });
  it('can accept a color array in options', () => {
     const black = new Color({red: 0, green: 0, blue: 0});
     const colors: Color[] = [
      black,
      black,
      black,
      black,
      black,
      black,
      black,
      black,
      black,
      black,
      black,
      black,
      black,
      black,
      black,
      black,
    ];
     const palette = new Palette({ name: 'pallete', colors });
     expect(palette.colors).toEqual(colors);
  });
  it('can accept a color array of varying lengths in options', () => {
    const black = new Color({red: 0, green: 0, blue: 0});
    const white =  new Color({red: 255, green: 255, blue: 255});
    const colors: Color[] = [
     black,
     black,
     black,
     black,
     black,
     black,
     black,
   ];
    const expectedColors: Color[] = [
    black,
    black,
    black,
    black,
    black,
    black,
    black,
    white,
    white,
    white,
    white,
    white,
    white,
    white,
    white,
    white,
  ];
    const palette = new Palette({ name: 'pallete', colors });
    expect(palette.colors).toEqual(colors);
 });
  it('gets an id', () => {
    const palette = new Palette({
      name: '',
    });
    expect(palette.id).toBeTruthy();
  });
  it('has a color array of the correct length of White', () => {
    const palette = new Palette({name: ''});
    expect(palette.colors).toMatchSnapshot();
  });
  it('updates valid index', () => {
    const palette = new Palette({name: ''});
    palette.updateColor(3, new Color({red: 0, green: 0, blue: 0}));
    expect(palette.colors).toMatchSnapshot();
  });
  it('rejects invalid index', () => {
    const palette = new Palette({name: ''});
    palette.updateColor(-1, new Color({red: 0, green: 0, blue: 0}));
    palette.updateColor(23, new Color({red: 0, green: 0, blue: 0}));
    expect(palette.colors).toMatchSnapshot();
  });
  it('can be duplicated', () => {
    const origPalette = new Palette({name: 'Original Palette'});
    const copyPalette = Palette.duplicate(origPalette);
    expect(copyPalette.name).toBe(origPalette.name);
    expect(copyPalette.id).toBeTruthy();
    expect(copyPalette.id).toEqual(expect.not.stringMatching(origPalette.id));
  });
  it('can update it\'s name', () => {
    const palette = new Palette({});
    const name = 'badgerville';
    palette.name = name;
    expect(palette.name).toEqual(name);
  });
});

