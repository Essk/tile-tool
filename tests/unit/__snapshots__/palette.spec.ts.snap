// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Palette has a color array of the correct length of White 1`] = `
Array [
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
]
`;

exports[`Palette rejects invalid index 1`] = `
Array [
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
]
`;

exports[`Palette updates valid index 1`] = `
Array [
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 0,
    "_green": 0,
    "_hex": "#000000",
    "_red": 0,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
  Color {
    "_blue": 255,
    "_green": 255,
    "_hex": "#FFFFFF",
    "_red": 255,
    "threeBitScale": Array [
      0,
      36,
      73,
      109,
      146,
      182,
      219,
      255,
    ],
    "twoBitScale": Array [
      0,
      85,
      170,
      255,
    ],
  },
]
`;
